client/asset/xmr/toolsdl/download.go

```diff
@@ -0,0 +1,869 @@
+	dirs, err := d.getAllLocalToolsDirs()
+	if err != nil {
+		return false, "", "", err
+	}
+	numDirs := len(dirs)
+
+	if numDirs == 0 {
+		return false, "", "", nil
+	}
+
+	if numDirs == 1 {
+		return true, filepath.Join(d.getToolsBasePath(), dirs[0].Name()), dirs[0].Name(), nil
+	}
+
+	// more than one tools dir
+
+	// find the one dir that is most up to date
+	type latestVer struct {
+		mv      *moneroVersionV0
+		dirname string
+	}
+	latest := &latestVer{
+		mv:      moneroVersionZeroV0(), // v0.0.0.0
+		dirname: "",
+	}
+
+	laterVersionFound := false
+	for _, dentry := range dirs {
+		mv, err := newMoneroVersionFromDir(dentry.Name())
+		if err != nil {
+			d.Log.Warnf("bad dir name: %s seen in GetBestCurrentLocalToolsDir", dentry.Name())
+			continue
+		}
+		rc := mv.compare(latest.mv)
+		if rc >= 0 {
+			// this entry version is >= latest looked at
+			latest.dirname = dentry.Name()
+			latest.mv = mv
+			laterVersionFound = true
+		}
+	}
+	if laterVersionFound {
+		return true, filepath.Join(d.getToolsBasePath(), latest.dirname), latest.dirname, nil
+	}
+
+	return false, "", "", nil
+}
+
+func (d *Download) getAllLocalToolsDirs() ([]os.DirEntry, error) {
+	toolsPath := d.getToolsBasePath()
+	// does toolsPath exist
+	_, err := os.Stat(toolsPath)
+	if err != nil {
+		return nil, nil // not an error
+	}
+	entries, err := os.ReadDir(toolsPath)
+	if err != nil {
+		return nil, fmt.Errorf("error reading tools dir - %w", err)
+	}
+
+	var dirs []os.DirEntry
+	for _, entry := range entries {
+		if entry.IsDir() && strings.HasPrefix(entry.Name(), "monero") {
+			hasBins, err := hasMoneroBinaries(filepath.Join(toolsPath, entry.Name()))
+			if err != nil {
+				return nil, err
+			}
+			if hasBins {
+				dirs = append(dirs, entry)
+			}
+		}
+	}
+	return dirs, nil
+}
+
+func hasMoneroBinaries(dirPath string) (bool, error) {
+	var gotCli, gotRpc bool
+	entries, err := os.ReadDir(dirPath)
+	if err != nil {
+		return false, err
+	}
+	for _, entry := range entries {
+		switch entry.Name() {
+		case "monero-wallet-cli", "monero-wallet-cli.exe":
+			gotCli = true
+		case "monero-wallet-rpc", "monero-wallet-rpc.exe":
+			gotRpc = true
+		}
+	}
+	return gotCli && gotRpc, nil
+}
+
+// ============================== remote ======================================
+
+var ErrNoRemoteVersion = errors.New("cannot find appropriate tools version in hashes.txt")
+
+// downloadRemoteCanonicalHashesFile downloads hashes.txt from getmonero.org and stores it's
+// local filepath.
+func (d *Download) downloadRemoteCanonicalHashesFile() {
+	hashesFilePath, err := d.downloadHashesFile()
+	if err != nil {
+		return
+	}
+	d.Log.Trace("Hashes file downloaded")
+	err = d.getHashedZips(hashesFilePath)
+	if err != nil {
+		return
+	}
+	d.Log.Trace("Collected hashed zips")
+	if len(d.hzips) <= 0 {
+		return
+	}
+	err = d.checkHashedZips()
+	if err != nil {
+		return
+	}
+	d.Log.Trace("Checked hashed zips")
+	err = d.chooseHashedZip()
+	if err != nil {
+		return
+	}
```

Comment:
Please continue to log all errors. I might even do Warn here. We don't expect any of these to fail, and we should continue, but failure may indicate a problem.

So probably return the errors and the consumer can log.